[FX] Preserve type annotations on generated code in Graph

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45880 [FX] Preserve type annotations on generated code in Graph**

Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

Commit:
[FX] Preserve type annotations on generated code in Graph

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -202,12 +211,22 @@ def _register_name_used(self, op : str) -> str:
+        modules_used : Set[str] = set(['typing'])
         body: List[str] = []
+
+        def type_repr(o : Any):
```
@zdevito  I have pretty low confidence that this function is correct in all cases. Is there any generalized way you know of to go from a Python object to an expression that would represent that object?

dzhulgakov replied:
maybe this one would work? https://github.com/pytorch/pytorch/blob/1a2d3b6a750cc4c1709b64412619b28c737e1a79/torch/__init__.py#L231

zdevito replied:
I don't really know. There is no guarantees here because it can always be some custom defined user class that has a weird `__repr__` defined. Or it can be a class whose name does not match the name in the global module space.  You can test it though... for instance, iterate over all of `torch` `torch.nn.function`. Look at all functions and all methods of classes. assert that `eval(repr(x)) == x` and see what breaks.

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -66,9 +67,17 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
+        """
         self._nodes : List[Node] = []
         self._used_names : Dict[str, int] = {}  # base name -> number
+        self.annotations = annotations_dict if annotations_dict else {}
```
why is it attached on the graph level and not a field on Node level? Arguments for the latter:
* easier to preserve types if I remove/rename the inputs (as the annotation lives on the placeholder node directly)
* may be convenient to preserve annotations on some intermediate nodes, e.g. if I happen to inline and un-inline some method and want to restore the annotations
* I'd argue that we might even want to capture output annotations when some module is called (just in case)

jamesr66a replied:
Originally I was just going to attach this dict as `__annotations__` on the generated `forward()` function, but it turns out that doesn't work so I ended up generating the type annotations as code. Agree that after that change this API doesn't make much sense (for my reasons and your reasons)

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Makes the `__annotations__` dict a parameter of the `Graph` constructor. `Graph` uses it to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in an annotations dict with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

jamesr66a commented on torch/__init__.py:
```diff
@@ -245,6 +245,15 @@ def typename(o):
+    # TODO: There's gotta be a better way to do this
+    if module == 'typing.':
+        return str(o)
+
```
@dzhulgakov @zdevito this seems sketchy to me since it's too permissive in what things in `typing` it matches on, but I didn't find a better way to check "just type aliases". On my python version the type aliases subclass `typing._GenericAlias` but that doesn't seem to be consistent. Any ideas if there's a better way.

dzhulgakov replied:
I think the problem is that `typename(o)` returns the type of the given object. Here you want something that gives a str representation of something that is already a type. Since types are objects too it gets confusing as usages within pytorch codebase are mixed, I see both `typename(obj)` and `typename(type(obj))`.

You probably should create a separate function (sorry for confusion) but you can follow the logic similar to here to handle builtins and stuff.

jamesr66a replied:
> I think the problem is that typename(o) returns the type of the given object. 

This is only true for the codepath on line 246, all the other codepaths assume that `o` is a type itself. The contract for this function seems to be "given an object, return the name of its type. given a type, return the qualified name of that type"

dzhulgakov replied:
Yeah, it's messed up :) There's https://github.com/python/cpython/blob/master/Lib/typing.py#L156 also but it seems similar to what goes on here.

jamesr66a replied:
k i'm gonna copy and paste that so i have plausible deniability if something is busted

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -121,16 +121,20 @@ def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> boo
+        args.extend(make_proxy_placeholder() for _ in range(1, total_args))
 
         if co.co_kwonlyargcount > 0 or co.co_flags & HAS_VARSTUFF:
             if co.co_flags & inspect.CO_VARARGS:
```
does python allow annotating `*args`? I'm not sure why one'd do that, but maybe to specify that all `*args` are the same class type or something like it. It'd require some tricks, because the type annotation needs to be applied to every unpacked element instead of the entire tuple (https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs). Feel free to skip it for now and just add a todo

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -313,20 +337,26 @@ def format_arg(arg) -> str:
             elif n.op == 'get_attr':
-                return f'%{n.name} : [uses={n.uses}] = self.{n.target}'
+                maybe_typename = f', type={torch.typename(n.type)}' if n.type is not None else ''
+                return f'%{n.name} : [uses={n.uses}{maybe_typename}] = self.{n.target}'
```
nit: maybe follow Python's annotation syntax and keep `[]` only for uses?

E.g. `foo : int [uses=2] = 123` instead of `foo : [uses=2, type=int] = 123` now?

dzhulgakov reviewed (commented):
As a node attribute it looks much nicer! Looks good overall, but @zdevito probably knows more about python typing tricks

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

dzhulgakov commented on torch/__init__.py:
```diff
@@ -245,6 +245,15 @@ def typename(o):
+    #
+    # TODO: There's gotta be a better way to do this
+    if module == 'typing.':
+        return str(o)
```
nit: it probably should be `repr` and not `str` (even if they are the same)

dzhulgakov reviewed (approved):
Should be good, we can always tweak it later if we encounter problems

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]

Commit:
Update on "[FX] Preserve type annotations on generated code in Graph"


Simple patch to address https://github.com/pytorch/pytorch/issues/45220. Attaches parameter and return value type annotations to `placeholder` and `output` nodes, respectively. `Graph` uses them to generate the proper function signature line during `python_code()`. Anyone generating a Graph programmatically can also pass in annotations with the proper type values. Currently leaves determining the proper type values up to the pass writer

Differential Revision: [D24127303](https://our.internmc.facebook.com/intern/diff/D24127303)

[ghstack-poisoned]
